File: log_search.cpp
// log_search.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include <algorithm>
#include <cctype>
#include <getopt.h>

using namespace std;

string highlight(const string& line, const string& pattern, bool regex) {
    if (regex) {
        regex re(pattern);
        return regex_replace(line, re, "\033[33m$&\033[0m");
    }
    string result = line;
    size_t pos = 0;
    while ((pos = result.find(pattern, pos)) != string::npos) {
        result.replace(pos, pattern.length(), "\033[33m" + pattern + "\033[0m");
        pos += pattern.length() + 11; // length of color codes
    }
    return result;
}

void searchFiles(const vector<string>& files, const string& pattern, bool regex, bool ignoreCase,
                 int context, bool lineNumbers, bool countOnly, const string& outputFile) {
    int totalMatches = 0;
    vector<string> results;
    regex compiled;
    bool compiledOk = false;
    if (regex) {
        try {
            compiled = regex(pattern, ignoreCase ? regex::icase : regex::ECMAScript);
            compiledOk = true;
        } catch (const regex_error& e) {
            cerr << "Regex error: " << e.what() << endl;
            return;
        }
    }

    for (const auto& filepath : files) {
        ifstream f(filepath);
        if (!f.is_open()) {
            cerr << "File not found: " << filepath << endl;
            continue;
        }
        vector<string> lines;
        string line;
        while (getline(f, line)) {
            lines.push_back(line);
        }
        f.close();

        if (countOnly) {
            int matchCount = 0;
            for (const auto& l : lines) {
                if (regex) {
                    if (regex_search(l, compiled)) matchCount++;
                } else {
                    if (ignoreCase) {
                        string lc = l, pc = pattern;
                        transform(lc.begin(), lc.end(), lc.begin(), ::tolower);
                        transform(pc.begin(), pc.end(), pc.begin(), ::tolower);
                        if (lc.find(pc) != string::npos) matchCount++;
                    } else {
                        if (l.find(pattern) != string::npos) matchCount++;
                    }
                }
            }
            totalMatches += matchCount;
            results.push_back(filepath + ": " + to_string(matchCount));
            continue;
        }

        vector<int> matchedIndices;
        for (size_t i = 0; i < lines.size(); i++) {
            const auto& l = lines[i];
            if (regex) {
                if (regex_search(l, compiled)) matchedIndices.push_back(i);
            } else {
                if (ignoreCase) {
                    string lc = l, pc = pattern;
                    transform(lc.begin(), lc.end(), lc.begin(), ::tolower);
                    transform(pc.begin(), pc.end(), pc.begin(), ::tolower);
                    if (lc.find(pc) != string::npos) matchedIndices.push_back(i);
                } else {
                    if (l.find(pattern) != string::npos) matchedIndices.push_back(i);
                }
            }
        }

        if (context > 0) {
            vector<int> shown;
            for (int idx : matchedIndices) {
                for (int offset = -context; offset <= context; offset++) {
                    int n = idx + offset;
                    if (n >= 0 && n < (int)lines.size()) shown.push_back(n);
                }
            }
            sort(shown.begin(), shown.end());
            shown.erase(unique(shown.begin(), shown.end()), shown.end());
            for (int idx : shown) {
                string prefix = (find(matchedIndices.begin(), matchedIndices.end(), idx) != matchedIndices.end()) ? ">" : " ";
                string line = lines[idx];
                if (prefix == ">") line = highlight(line, pattern, regex);
                string lineNum = lineNumbers ? to_string(idx+1) : "";
                if (!lineNum.empty()) lineNum = string(4 - min(4, (int)lineNum.length()), ' ') + lineNum;
                results.push_back(filepath + ":" + lineNum + ": " + line);
            }
            totalMatches += matchedIndices.size();
        } else {
            for (int idx : matchedIndices) {
                string line = lines[idx];
                line = highlight(line, pattern, regex);
                string lineNum = lineNumbers ? to_string(idx+1) : "";
                if (!lineNum.empty()) lineNum = string(4 - min(4, (int)lineNum.length()), ' ') + lineNum;
                results.push_back(filepath + ":" + lineNum + ": " + line);
            }
            totalMatches += matchedIndices.size();
        }
    }

    if (!outputFile.empty()) {
        ofstream out(outputFile);
        for (const auto& r : results) out << r << "\n";
        if (!countOnly) out << "\nTotal matches: " << totalMatches << "\n";
        out.close();
        cout << "Results written to " << outputFile << "\n";
    } else {
        for (const auto& r : results) cout << r << "\n";
        if (!countOnly) cout << "\nTotal matches: " << totalMatches << "\n";
    }
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        cerr << "Usage: log_search <pattern> <file1> [file2...] [options]\n";
        cerr << "Options: --regex, --ignore-case, --context N, --line-numbers, --count, --output FILE\n";
        return 1;
    }

    string pattern = argv[1];
    vector<string> files;
    bool regex = false, ignoreCase = false, lineNumbers = false, countOnly = false;
    int context = 0;
    string outputFile;

    for (int i = 2; i < argc; i++) {
        string arg = argv[i];
        if (arg == "--regex") regex = true;
        else if (arg == "--ignore-case") ignoreCase = true;
        else if (arg == "--line-numbers") lineNumbers = true;
        else if (arg == "--count") countOnly = true;
        else if (arg == "--context" && i+1 < argc) context = stoi(argv[++i]);
        else if (arg == "--output" && i+1 < argc) outputFile = argv[++i];
        else files.push_back(arg);
    }

    if (files.empty()) {
        cerr << "No files specified.\n";
        return 1;
    }

    searchFiles(files, pattern, regex, ignoreCase, context, lineNumbers, countOnly, outputFile);
    return 0;
}

File: log_search.php
# log_search.php
#!/usr/bin/env php
<?php

function highlight($line, $pattern, $regex) {
    if ($regex) {
        return preg_replace('/' . preg_quote($pattern) . '/', "\033[33m\\0\033[0m", $line);
    }
    return str_replace($pattern, "\033[33m{$pattern}\033[0m", $line);
}

function searchFiles($files, $pattern, $regex, $ignoreCase, $context, $lineNumbers, $countOnly, $output) {
    $totalMatches = 0;
    $results = [];

    foreach ($files as $filepath) {
        if (!file_exists($filepath)) {
            fwrite(STDERR, "File not found: $filepath\n");
            continue;
        }
        $lines = file($filepath, FILE_IGNORE_NEW_LINES);
        if ($lines === false) continue;

        if ($countOnly) {
            $matchCount = 0;
            foreach ($lines as $line) {
                if ($regex) {
                    if (preg_match('/' . preg_quote($pattern) . '/', $line)) $matchCount++;
                } else {
                    if ($ignoreCase) {
                        if (stripos($line, $pattern) !== false) $matchCount++;
                    } else {
                        if (strpos($line, $pattern) !== false) $matchCount++;
                    }
                }
            }
            $totalMatches += $matchCount;
            $results[] = "$filepath: $matchCount";
            continue;
        }

        $matchedIndices = [];
        foreach ($lines as $idx => $line) {
            if ($regex) {
                if (preg_match('/' . preg_quote($pattern) . '/', $line)) $matchedIndices[] = $idx;
            } else {
                if ($ignoreCase) {
                    if (stripos($line, $pattern) !== false) $matchedIndices[] = $idx;
                } else {
                    if (strpos($line, $pattern) !== false) $matchedIndices[] = $idx;
                }
            }
        }

        if ($context > 0) {
            $shown = [];
            foreach ($matchedIndices as $idx) {
                for ($offset = -$context; $offset <= $context; $offset++) {
                    $n = $idx + $offset;
                    if ($n >= 0 && $n < count($lines)) $shown[] = $n;
                }
            }
            $shown = array_unique($shown);
            sort($shown);
            foreach ($shown as $idx) {
                $prefix = in_array($idx, $matchedIndices) ? '>' : ' ';
                $line = $lines[$idx];
                if ($prefix === '>') $line = highlight($line, $pattern, $regex);
                $lineNum = $lineNumbers ? str_pad($idx+1, 4, ' ', STR_PAD_LEFT) : '';
                $results[] = "$filepath:$lineNum: $line";
            }
            $totalMatches += count($matchedIndices);
        } else {
            foreach ($matchedIndices as $idx) {
                $line = $lines[$idx];
                $line = highlight($line, $pattern, $regex);
                $lineNum = $lineNumbers ? str_pad($idx+1, 4, ' ', STR_PAD_LEFT) : '';
                $results[] = "$filepath:$lineNum: $line";
            }
            $totalMatches += count($matchedIndices);
        }
    }

    if ($output) {
        file_put_contents($output, implode("\n", $results) . "\n");
        if (!$countOnly) file_put_contents($output, "\nTotal matches: $totalMatches\n", FILE_APPEND);
        echo "Results written to $output\n";
    } else {
        echo implode("\n", $results) . "\n";
        if (!$countOnly) echo "\nTotal matches: $totalMatches\n";
    }
}

$opts = getopt("", ["regex", "ignore-case", "context:", "line-numbers", "count", "output:"]);
$args = array_slice($argv, 1);
foreach ($args as $i => $arg) {
    if (strpos($arg, '--') === 0) unset($args[$i]);
}
$args = array_values($args);
if (count($args) < 2) {
    echo "Usage: php log_search.php <pattern> <file1> [file2...] [options]\n";
    exit(1);
}
$pattern = array_shift($args);
$files = $args;
$regex = isset($opts['regex']);
$ignoreCase = isset($opts['ignore-case']);
$context = isset($opts['context']) ? (int)$opts['context'] : 0;
$lineNumbers = isset($opts['line-numbers']);
$countOnly = isset($opts['count']);
$output = $opts['output'] ?? null;

searchFiles($files, $pattern, $regex, $ignoreCase, $context, $lineNumbers, $countOnly, $output);
?>
